python/rmm/_lib/device_buffer.pyx

```diff
@@ -139,3 +159,28 @@ cdef class DeviceBuffer:
 
     cdef void* c_data(self):
         return self.c_obj.get()[0].data()
+
+
+@cython.boundscheck(False)
+cpdef void copy_to_host(uintptr_t db,
+                        unsigned char[::1] hb,
+                        uintptr_t stream=0) nogil except *:
+    if hb is None:
+        with gil:
+            raise TypeError(
+                "Argument `hb` has incorrect type"
+                " (expected bytes-like, got NoneType)"
+            )
+
+    cdef cudaError_t err
+
+    err = cudaMemcpyAsync(<void*>&hb[0], <const void*>db, len(hb),
+                          cudaMemcpyDeviceToHost, <cudaStream_t>stream)
+    if err != cudaSuccess:
+        with gil:
+            raise RuntimeError(f"Memcpy failed with error: {err}")
+
+    err = cudaStreamSynchronize(<cudaStream_t>stream)
```

Comment:
Should there be an explicit synchronization here always? I think it could be useful to have a truly async version and let the user choose when to synchronize. I understand we want a safe version, but the user could for example enqueue several consecutive copies and only sync after the last one, thus avoiding the blocking behavior, plus, it seems that the C++ `copy_to_host` implementation used to behave like that.

Reply:
This basically matches the behavior of the `tobytes` implementation before.

That said, I'm not sure. Am a little wary to add that fearing users might shoot themselves in the foot. Besides it is happening while the GIL is released. So Python can actually still do other things while this happens in the background (like when using `async` with threads).

However I was curious what Numba is doing (as we want to match them) and CuPy (for another opinion). It seems they don't call `cudaStreamSynchronize` when a stream is provided. Though they do run synchronously if no stream is provided. Perhaps we should do the same? @kkraus14, do you have thoughts here?

http://numba.pydata.org/numba-doc/latest/cuda/memory.html#copy_to_host
https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.asnumpy.html#cupy.asnumpy

Reply:
Unfortunately, I'm not very familiar with how the GIL is supposed to work in such conditions, so I won't comment on that.

> Am a little wary to add that fearing users might shoot themselves in the foot. [...] However I was curious what Numba is doing (as we want to match them) and CuPy (for another opinion). It seems they don't call cudaStreamSynchronize when a stream is provided.

Yeah, this is the way CUDA generally is supposed to work. If no stream is provided, the default stream is assumed and, thus, synchronous behavior. However, if a stream is passed, then it's assumed that the user will take care of synchronization. With that said, I'm not necessarily saying we should always have an asynchronous behavior, but it could be useful to have a "safe" version (what you have), and another one where the user needs to ensure synchronization.

Reply:
If a user uses the `nvcc` option for per thread default stream and we synchronize here what's the behavior? Normally if you synchronize on the default stream it basically synchronizes the entire device which is something we definitely don't want to do.

Reply:
> If a user uses the `nvcc` option for per thread default stream and we synchronize here what's the behavior? Normally if you synchronize on the default stream it basically synchronizes the entire device which is something we definitely don't want to do.

I'm guessing this is a question for @pentschev? Am not sure myself 🙂

<hr>

In any event, I'm happy to adjust the behavior based on our discussion.

Think I may know the answer to this question, but want to confirm. Does `cudaMemcpyAsync` synchronize if `stream` is `0`? Or do we need to do that ourselves?

Reply:
> If a user uses the `nvcc` option for per thread default stream and we synchronize here what's the behavior? Normally if you synchronize on the default stream it basically synchronizes the entire device which is something we definitely don't want to do.

Looks like each thread gets its own default stream. Turns out Mark wrote [a blog about this]( https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/ ) 😄

Reply:
Have pushed a commit to drop the sync step. Please let me know if I need to do anything else here.

Reply:
We changed this to synchronize only on the default stream, but no others. Starting [a new thread below]( https://github.com/rapidsai/rmm/pull/268#discussion_r373292883 ) for further discussion as needed.